compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/SingleAbstractMethodLowering.kt

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.backend.jvm.lower
+
+import org.jetbrains.kotlin.backend.common.*
+import org.jetbrains.kotlin.backend.common.ir.copyTo
+import org.jetbrains.kotlin.backend.common.ir.createImplicitParameterDeclarationWithWrappedDescriptor
+import org.jetbrains.kotlin.backend.common.lower.createIrBuilder
+import org.jetbrains.kotlin.backend.common.lower.irBlock
+import org.jetbrains.kotlin.backend.common.phaser.makeIrFilePhase
+import org.jetbrains.kotlin.backend.common.serialization.fqNameSafe
+import org.jetbrains.kotlin.backend.jvm.JvmLoweredDeclarationOrigin
+import org.jetbrains.kotlin.descriptors.ClassKind
+import org.jetbrains.kotlin.descriptors.Modality
+import org.jetbrains.kotlin.descriptors.Visibilities
+import org.jetbrains.kotlin.ir.builders.*
+import org.jetbrains.kotlin.ir.builders.declarations.*
+import org.jetbrains.kotlin.ir.declarations.*
+import org.jetbrains.kotlin.ir.expressions.*
+import org.jetbrains.kotlin.ir.expressions.impl.IrInstanceInitializerCallImpl
+import org.jetbrains.kotlin.ir.types.IrType
+import org.jetbrains.kotlin.ir.types.classifierOrFail
+import org.jetbrains.kotlin.ir.types.toKotlinType
+import org.jetbrains.kotlin.ir.util.*
+import org.jetbrains.kotlin.ir.visitors.transformChildrenVoid
+import org.jetbrains.kotlin.load.java.sam.SamConstructorDescriptor
+import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.types.KotlinType
+
+internal val singleAbstractMethodPhase = makeIrFilePhase(
+    ::SingleAbstractMethodLowering,
+    name = "SingleAbstractMethod",
+    description = "Replace SAM conversions with instances of interface-implementing classes"
+)
+
+class SingleAbstractMethodLowering(val context: CommonBackendContext) : ClassLoweringPass {
+    override fun lower(irClass: IrClass) {
+        // TODO use `IrType`s; unfortunately, they are currently not comparable.
+        //      (Can't use class symbols here, we want type parameters as well.)
+        val cachedImplementations = mutableMapOf<Pair<KotlinType, KotlinType>, IrClass>()
+        val localImplementations = mutableListOf<IrClass>()
+        irClass.transformChildrenVoid(object : IrElementTransformerVoidWithContext() {
+            override fun visitCall(expression: IrCall): IrExpression {
+                if (expression.symbol.owner.descriptor !is SamConstructorDescriptor)
```

Comment:
@dnpetrov Do we have, or can we add a descriptor-less way to detect that `IrFunction` is a SAM constructor? Maybe a new origin or something?

Reply:
Ideally, SamConstructorDescriptor should not be called in IR generated by psi2ir: SAM conversion and explicit SAM construction becomes IrTypeOperatorCall for IrTypeOperator.SAM_CONVERSION. 
If SamConstructorDescriptor calls "leak" to IR somehow, it's probably a psi2ir bug or underdesign on my side.

Reply:
Must be a bug; this seems to happen when the interface has type parameters, in which case the unbound descriptor is a SamConstructorDescriptor, but the one with substituted parameters is a SimpleFunctionDescriptor, so the call slips by the check in psi2ir.

Reply:
Added a fix and a test. The additional IMPLICIT_CAST to `kotlin.Function2<T of java.util.Comparator?, T of java.util.Comparator?, kotlin.Int>` looks suspicious to me, though.